include/rsp_dma.inc

```diff
@@ -0,0 +1,183 @@
+#   * DMAOutAsync: start a background DMEM -> RDRAM transfer
+#   * DMAExec: single entry point for any type of transfer. It can be used
+#     to write direction-agnostic transfer code that can be then invoked
+#     passing in the direction in a register.
+#
+# The RSP DMA controller allows to enqueue one transfer while another one
+# is in progress. You can use the *Async functions to do this. If you try
+# to enqueue a third transfer, though, the *Async function will automatically
+# block waiting for the first one to finish. This allows the code to call
+# *Async functions freely and get "best effort" background transfers.
+# 
+# When you need to explicitly synchronize (eg: to access read data), you can
+# use DMAWaitReady to wait until the DMA engine is ready to receive another
+# transfer, or DMAWaitIdle to wait until all DMA transfers are finished. See
+# their documentation below.
+#
+# The DMA controller is able to read/write "rectangular portions"
+# of memory in RDRAM (that is, you can define width, height and
+# pitch to extract a portion of a larger image). Accesses in DMEM
+# are always linear, though. If you don't need this feature,
+# you can set the transfer size as width and use 1 as height: this way,
+# the pitch will be ignored.
+#
+# At the hardware level, both the DMEM and the RDRAM addresses are
+# treated as 8-byte aligned (that is, the lowest 3 bits are ignored).
+# If you issue a transfer from RDRAM with an unaligned address (in s0),
+# the DMEM address (s4) will be updated to reflect where the byte that
+# was actually requested ended up. For instance, if s0 = 0x80041203, s4=0x100,
+# after the function s4 will be 0x103, as that is the address where the
+# byte at 0x80041203 ended up being written to.
+#
+# INPUT:
+#   t0: transfer size as width/height (use DMA_SIZE macro)
+#   t1: pitch (ignored if height==1)
+#   s0: RDRAM address
+#   s4: DMEM address. After DMA, updated with unalignment
+#
+#   Only for DMAExec:
+#   t2: DMA operation flags: select direction and sync/async.
+#       Use one of the DMA_* macros below. If async, you can call
+#       DMAWaitIdle to synchronize later if required.
+#
+# OUTPUT:
+#   Only for DMAIn:
+#   s4: adjusted to point to the first requested byte
+#       (because of DMA alignment, more bytes might have been
+#       DMA'd).
+#
+# DESTROY:
+#   t2, at
+# 
+#########################################################
+
+    .func DMAOut
+DMAOut:
+    j DMAExec
+    li t2, DMA_OUT
+    .endfunc
+
+    .func DMAOutAsync
+DMAOutAsync:
+    j DMAExec
+    li t2, DMA_OUT_ASYNC
+    .endfunc
+
+    .func DMAInAsync
+DMAInAsync:
+    j DMAExec
+    li t2, DMA_IN_ASYNC
+    .endfunc
+
+    .func DMAIn
+DMAIn:
+    li t2, DMA_IN
+    # fallthrough!
+    .endfunc
+
+    .set noat
+    .func DMAExec
+DMAExec:
+    # Wait for the DMA engine to be ready. We do this
+    # every time because with async transfers it's easy to get
+    # confused and forget to wait.
```

Comment:
It can be added as an #ifdef of course, if required. 

In my experience trying to squeeze times, it's an important safety net. Code-wise, consider that if you remove this, you might be forced to add many calls to `DMAWaitReady` instead, which takes at least 1 instruction of space in IMEM per each site.